### ISSUE TITLE
[TD-67] "Could not find common ancestor" error

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskserver#68]
_Jack on 2014-07-27T04:32:26Z says:_ 

 I got this error:

{code}
2014-07-27 04:26:09 [53] 'sync' from 'FreeCinc/freecinc_ea43a73a' using 'Mirakel 2.6.2' at :0
2014-07-27 04:26:09 [53] Client key '' 41
2014-07-27 04:26:09 [53] Loaded 45 records
2014-07-27 04:26:09 [53] Subset 43 tasks
2014-07-27 04:26:09 [53] patch add modified=1406433877
2014-07-27 04:26:09 [53] Merge result {"description":"rub out the numbers on my desk thermometer and replace them with a comfort zone with non-labeled degree markings","entry":"20140702T184551Z","modified":"20140727T040437Z","status":"pending","uuid":"82deb6be-1267-4491-9b71-bdf4c6924786"}
2014-07-27 04:26:09 [53] ERROR: Could not find common ancestor for 0dd22582-7c42-4b24-b46e-d453f252c7bc
2014-07-27 04:26:09 [53] Serviced in 0.229839s
{code}

Basically I had configured both TaskWarrior and Mirakel to sync with one account, then I reconfigured both to sync with a different account (different server, even). I added one more task from the TaskWarrior side, synced TaskWarrior (Along the way I tried to sync TaskWarrior without explicitly adding `init` to my `task sync` call). Then when I went to sync Mirakel, this is the error I saw in the server logs.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2014-07-27T04:32:26Z
Modified: 2015-06-22T15:04:02Z
```

**Comment by taskwarrior** [comment]
_Jack on 2014-08-02T04:38:40Z says:_ 

 I'm getting this error again. I wasn't sure how to fix it, so I deleted all the *.data files from ~/.task and then re-added the tasks.

**Comment by taskwarrior** [comment]
_Renato Alves on 2014-08-13T14:16:55Z says:_ 

 Hi Jack,

I'll try to reproduce this problem using only taskwarrior clients. There's not much I can do about Mirakel.
Perhaps you should report this to the Mirakel team too, or link this bug on their bugtracker.

**Comment by taskwarrior** [comment]
_Jack on 2014-08-13T16:17:29Z says:_ 

 Hi Renato,

Here is a link to the corresponding Mirakel bug: https://github.com/MirakelX/mirakel-android/issues/456

**Comment by taskwarrior** [comment]
_Renato Alves on 2014-10-16T11:35:05Z says:_ 

 Hi Jack,

I've started using Mirakel myself and was already bit by this bug.
It happened not on first sync but after importing a large set of tasks.
The offending task however was a recurrent one.

I've yet to reproduce it with taskwarrior clients alone but now I have some ideas of where to look.

**Comment by taskwarrior** [comment]
_Renato Alves on 2014-10-16T11:56:58Z says:_ 

 Can you check if the UUID 0dd22582-7c42-4b24-b46e-d453f252c7bc is a children or parent of a recurring task?

**Comment by taskwarrior** [comment]
_Jack on 2014-10-16T16:09:19Z says:_ 

 I'm not that familiar with parent/child relationships in taswarrior, so I've pasted in the entire task as seen by `task edit`. (This happens to be one of my tasks, so it's not obfuscated at all)

Does ID == 1 have any significance here?

{code}
1.  Name               Editable details
1.  -----------------  ----------------------------------------------------
1.  ID:                1
1.  UUID:              0dd22582-7c42-4b24-b46e-d453f252c7bc
1.  Status:            Pending
1.  Mask:
1.  iMask:
  Project:           pizzazz
  Priority:
1.  Separate the tags with spaces, like this: tag1 tag2
  Tags:
  Description:       teach daily_lager to find things faster with indices
  Created:           2014-07-07 04:09:17
  Started:
  Ended:
  Scheduled:
  Due:
  Until:
  Recur:
  Wait until:
  Parent:
1.  Annotations look like this: <date> -- <text> and there can be any number of them.
1.  The ' -- ' separator between the date and text field should not be removed.
1.  A "blank slot" for adding an annotation follows for your convenience.
  Annotation:        2014-10-16 16:04:43 --
1.  Dependencies should be a comma-separated list of task IDs/UUIDs or ID ranges, with no spaces.
  Dependencies:
1.  User Defined Attribute Orphans
  UDA Orphan progress: 53.000000
1.  End
{code}

**Comment by taskwarrior** [comment]
_Renato Alves on 2014-10-16T17:04:09Z says:_ 

 This doesn't seem to be a recurring task (parent is not defined).
Having ID=1 only means it's the first task in your pending.data file.

Would you mind also checking if UUID 0dd22582-7c42-4b24-b46e-d453f252c7bc exists on the tx.data file matching your taskd account?

Following your question about ID=1, have you ever synced your task database to a different taskd account in the (recent) past?

**Comment by taskwarrior** [comment]
_Jack on 2014-10-17T15:50:30Z says:_ 

 Yes, it does exist in the tx.data file:

{code}
$ cat tx.data | grep 0dd22582-7c42-4b24-b46e-d453f252c7bc
{"description":"teach daily_lager to find things faster with indices","entry":"20140707T040917Z","modified":"20140723T061519Z","project":"lager","status":"pending","uuid":"0dd22582-7c42-4b24-b46e-d453f252c7bc"}
{"uuid":"0dd22582-7c42-4b24-b46e-d453f252c7bc","status":"pending","entry":"20140707T040917Z","description":"teach daily_lager to find things faster with indices","project":"pizzazz","modified":"20140823T214549Z","progress":53}
{code}

**Comment by taskwarrior** [comment]
_Jack on 2014-10-17T15:53:03Z says:_ 

 I've probably synced my same set of tasks with about five different taskd accounts. It's probably been about a month or more since I changed to a new account with the same tasks.

**Comment by taskwarrior** [comment]
_ScrumpyJack on 2015-03-17T16:56:13Z says:_ 

 same sort of error here with taskwarrior


Syncing with freecinc.com:53589

Sync failed.  The Taskserver returned error: 500 ERROR: Could not find
common ancestor for 758a08bf-4e41-4128-8ae4-aea42fba54cc

I can provide a shell to a taskwarrior container if needed

**Comment by taskwarrior** [comment]
_Renato Alves on 2015-03-17T17:29:41Z says:_ 

 Hi ScrumpyJack,

Can you provide information regarding your steps in getting to the point that taskwarrior returns the error, including answers to the following:
 * How many clients do you use? Are they all taskwarrior or other clients have been used (mirakel, etc.)?
 * Do you use `task undo`? Frequently?
 * Do you have recurring tasks? 
 * Can you provide the full information regarding `758a08bf-4e41-4128-8ae4-aea42fba54cc` (anonymized is fine)?
 * Have you ever manually copied your `.task` data folder between computers before/after using sync for the first time?

Anything else you can think of that may be relevant to reproduce the problem is also welcome.
Thanks

**Comment by taskwarrior** [comment]
_ScrumpyJack on 2015-06-22T13:11:32Z says:_ 

 How many clients do you use? 3
Do you use task undo? Frequently? Infrequently, but I have used it. 
Do you have recurring tasks? None
Have you ever manually copied your .task? Never

**Comment by taskwarrior** [comment]
_Renato Alves on 2015-06-22T15:04:02Z says:_ 

 Hi ScrumpyJack,

That's interesting.data. Recurring tasks are the number one source for problems so far.
You seem to reach this state without that. So other than the ocasional "undo" or a bad client, I don't really know how you got there.

Are these 3 clients all taskwarrior? No mirakel?

**Comment by BlitzKraft** [comment]
I recently encountered this issue using just taskwarrior, and a sync server. It is from a recurring task with virtual tags "child deleted unblocked project". And only one account. No undo used in the past year. task warrior version 2.5.1. 

Hope this helps in narrowing down the issue.

Edit: After reading the task docs ([this part](https://taskwarrior.org/docs/taskserver/troubleshooting-sync.html)) again, the resolution seemed to be  `task sync init`. In a way it is, but not without a few draw backs. The error popped up, for me, about a week may be two weeks ago. All the tasks in that duration were reset (status set to pending on both local and remote). For me, this meant manually cleaning up ~40 tasks. 

This is more of an observation, mainly because I can't confirm exactly when the error 500 started appearing. If it just started appearing, then it might suffice to `init`, again.

**Comment by PorcelainMouse** [comment]
I just rant into this.

Profile:
2 clients, linux & android (taskwc2)
many recurring tasks
offending task was not recurring
problem seemed to occur immediately after creating task. Unfortunately, I'm not sure which client I used to create it.  I'm pretty sure it was the Linux client.

Symptom:
Same as everyone else, but only the android app was throwing this error.  Linux client seemed to sync, fine.  Android client wasn't updating.

Solution:
I found the task in the backlog on the Android client.  I just deleted that line.  The next sync from the Android client worked.  After that, the Linux client pick up all the other changes that were waiting.  (Offending task was at top of backlog list.)  Offending task looks fine, no fall out.

**Comment by PorcelainMouse** [comment]
Ran into this again on my android client.  Same approach worked, i.e. searching for the offending task in all the .data files and removing it.  However, this time it was not in the backlog, but in both pending.data and completed.data.  FWIW.

**Comment by PorcelainMouse** [comment]
Ouch, okay, this finally caught up to me.  I had one task that got messed up and I couldn't fix it.  I occasionally get duplicates, don't know why, but this one was actually important because I was trying to keep careful track of my time.  So, I tried to fix it by manually deleting one of the duplicate tasks.  That resulted in this error.  I tried to fix that by deleting the backlog, and them entire history got modified during the next sync.  4 years of annotations go reset to just now.  And I have several task that came back from the dead, most are duplicates.  Bother.

I only sync with one android app and I use undo occasionally, but not often.  I wish task was more resilient, or fixable.

**Comment by cjadeveloper** [comment]
> _Jack on 2014-07-27T04:32:26Z says:_
> 
> I got this error:
> 
> {code}
> 2014-07-27 04:26:09 [53] 'sync' from 'FreeCinc/freecinc_ea43a73a' using 'Mirakel 2.6.2' at :0
> 2014-07-27 04:26:09 [53] Client key '' 41
> 2014-07-27 04:26:09 [53] Loaded 45 records
> 2014-07-27 04:26:09 [53] Subset 43 tasks
> 2014-07-27 04:26:09 [53] patch add modified=1406433877
> 2014-07-27 04:26:09 [53] Merge result {"description":"rub out the numbers on my desk thermometer and replace them with a comfort zone with non-labeled degree markings","entry":"20140702T184551Z","modified":"20140727T040437Z","status":"pending","uuid":"82deb6be-1267-4491-9b71-bdf4c6924786"}
> 2014-07-27 04:26:09 [53] ERROR: Could not find common ancestor for 0dd22582-7c42-4b24-b46e-d453f252c7bc
> 2014-07-27 04:26:09 [53] Serviced in 0.229839s
> {code}
> 
> Basically I had configured both TaskWarrior and Mirakel to sync with one account, then I reconfigured both to sync with a different account (different server, even). I added one more task from the TaskWarrior side, synced TaskWarrior (Along the way I tried to sync TaskWarrior without explicitly adding `init` to my `task sync` call). Then when I went to sync Mirakel, this is the error I saw in the server logs.

Try deleting `backlog.data` from `~/.task/` and sync again.

**Comment by Robinlovelace** [comment]
> Try deleting `backlog.data` from `~/.task/` and sync again.

In case another data point helps here, I tried tried that and then 

```bash
task sync
# but not task sync init
```

worked, many thanks for a great open source tool :tada:

**Comment by lauft** [comment]
> [!IMPORTANT]
> Taskserver is only compatible with Taskwarrior 2.x, and is no longer actively developed.
> See `man task-sync` for task synchronization with Taskwarrior 3
